### ISSUE TITLE
Cannot actually define multiple pillars with S3 ext_pillar multiple_env

### ISSUE THREAD

**Issue by EliRibble** [saltstack/salt#22471]
Currently specifying `multiple_env: True` on an ext_pillar using S3 will cause salt to download the pillars by environment but it will not cause it to fully create each of those environments. For example, let's say I specify the following in my master config:

```
ext_pillar:
  - s3:
    bucket: salt-pillars
    keyid: *redacted*
    key: *redacted*
    multple_env: True
    verify_ssl: True
```

And I have no pillar_roots in my config file. Salt will download the following files from my bucket

```
s3://salt-pillars/base/top.sls
s3://salt-pillars/prod/top.sls
```

However, `pillar/s3.py:ext_pillar` will return a compiled pillar where the environment is set to the default (`base`) that will not have merged in anything from `top.sls`. This means that for all intents and purposes `prod/top.sls` doesn't exist.

The correct behavior would be to return a compiled pillar per environment that is found but I have no idea how to do that since `ext_pillar` is supposed to return a single pillar rather than many.

One workaround that I have used is to specify my configuration as follows:

```
ext_pillar:
  - s3:
    bucket: salt-pillars
    keyid: *redacted*
    key: *redacted*
    multple_env: True
    verify_ssl: True
    environment: base
  - s3:
    bucket: salt-pillars
    keyid: *redacted*
    key: *redacted*
    multple_env: True
    verify_ssl: True
    environment: prod
```

In other words, you duplicate the configuration for each environment in your bucket. This means that salt will pull in your configurations N**2 times where N is the number of environments you've defined. This isn't so bad because of the caching that ext_pillar s3 uses.

**Comment by EliRibble** [comment]
See https://github.com/saltstack/salt/issues/22472 for a workaround

**Comment by rallytime** [comment]
Thanks for posting this report and for the workaround.

**Comment by rallytime** [comment]
@EliRibble With the change you made in #22484 help address this issue as well?

**Comment by EliRibble** [comment]
It doesn't really address the root issue unfortunately as the heart of the issue is that we can't, as far as I know, define multiple logical pillars in a single ext_pillar configuration.

**Comment by unirede** [comment]
Besides the workaround, another problem exists, the salt-master does not create a subdir and is need to create a directory manually. Errors in log:
```
2016-09-28 13:12:03,854 [salt.pillar      ][CRITICAL][15249] Pillar render error: Failed to load ext_pillar s3: [Errno 2] No such file or directory: '/var/cache/salt/master/pillar_s3fs/salt-pillars/prd-files.cache'
2016-09-28 13:12:03,855 [salt.pillar      ][CRITICAL][15249] Pillar render error: Failed to load ext_pillar s3: [Errno 2] No such file or directory: '/var/cache/salt/master/pillar_s3fs/salt-pillars/hml-files.cache'
2016-09-28 13:12:03,855 [salt.pillar      ][CRITICAL][15249] Pillar render error: Failed to load ext_pillar s3: [Errno 2] No such file or directory: '/var/cache/salt/master/pillar_s3fs/salt-pillars/dev-files.cache'

The config:
```
ext_pillar:
  - s3:
      bucket: salt
      keyid: xxxxx
      key: xxxxx
      environment: base
      prefix: pillars/prd
  - s3:
      bucket: salt
      keyid: xxxxx
      key: xxxxxx
      environment: hml
      prefix: pillars/hml
  - s3:
      bucket: salt
      keyid: xxxxx
      key: xxxxxx
      environment: dev
      prefix: pillars/dev
```
